lib/src/realm_class.dart

```diff
@@ -85,12 +86,20 @@ export 'session.dart' show Session, SessionState, ConnectionState, ProgressDirec
 ///
 /// {@category Realm}
 class Realm {
-  final Map<Type, RealmMetadata> _metadata = <Type, RealmMetadata>{};
-  final RealmHandle _handle;
+  late final RealmMetadata _metadata;
+  late final RealmHandle _handle;
+
+  /// An object encompassing this `Realm` instance's dynamic API.
+  late final DynamicRealm dynamic = DynamicRealm._(this);
```

Comment:
Why are we exposing this on the public interface. why not just have this as an argument to the migration function args. I think we should not make this public since there is no use case of this outside migrations. This is fully duplicating the typed interface so in a sense we are trying to make the SDK a dynamic one where we shouldn't. This SDK is for a typed language and should only provide typed language interface. At least we should try and in special cases like migrations we could provide a dynamic interface.

Reply:
I don't agree that there's a reason to limit the dynamic API to migrations. There are many use cases for the dynamic API outside of migrations:
1. Dart already exposes dynamic API despite being a statically typed language. So clearly there's some demand for it.
2. To write a Studio-like tool where you read the schema from disk.
3. To import dynamic data from an API (see [this forum post](https://www.mongodb.com/community/forums/t/how-to-add-dynamicrealmobject/119228) for such a request for the .NET SDK).
4. To handle unknown objects via sync. When we add support for Mixed/RealmValue, we can end up in a situation where an old version of the app is accessing objects that it doesn't have dart models for. Those need to be exposed as dynamic objects, otherwise the developer will have no way to interact with them.

Reply:
Our task at hand is migrations. We don't have any need/request to support the scenarios you listed. Mixed and RealmValue might end up implemented in a way that does not need what .NET has. I think we are doing too much implementation of anticipation of future features implemented in certain ways. 
Dart has `dynamic` keyword and dynamic objects but never in this PR we have this keyword mentioned. So in a sense we are not following the language idiomaticity. Maybe it is just  me but I don't really understand the scenarios we are trying to cover maybe because they don't exists yet. Perhaps we can discuss offline to better understand what we are trying to achieve here.